### PR TITLE
More const PoseGraph interface

### LINKED ISSUE #1021: PoseGraphInterface can mostly be const
Currently, all methods in `PoseGraphInterface` are not const, even though most of them can be.
I think the reason was that they require a read mutex, but semantically, they are const.

### PR DESCRIPTION
**Pull request by sebastianklose**
This PR makes getters in `PoseGraph*` `const`. 
Before most getters where marked non-const because of the need to acquire the mutex for thread safety. This forces all code using the PoseGraph to pass it around as a non-const object even if the consumer is not altering the object. 
By making the mutexes `mutable` (for details see [this talk by Herb Sutter](https://channel9.msdn.com/posts/C-and-Beyond-2012-Herb-Sutter-You-dont-know-blank-and-blank)) we can make getters and methods that do not change the PoseGraph const and thus enable consumer APIs to express how they are going to use the PoseGraph as well.

fixes #1021

### CODE REVIEW

**Comment by gaschler** [comment]
> CI says the PR does not compile.

Follow up changes in cartographer::cloud, for instance in PoseGraphStub, and likely in cartographer_ros, are needed.

File: cartographer/mapping/internal/connected_components.cc [hunk 1]
```diff
@@ -50,17 +50,17 @@ void ConnectedComponents::Union(const int trajectory_id_a,
   forest_[representative_a] = representative_b;
 }
 
-int ConnectedComponents::FindSet(const int trajectory_id) {
+int ConnectedComponents::FindSet(const int trajectory_id) const {
   auto it = forest_.find(trajectory_id);
   CHECK(it != forest_.end());
   if (it->first != it->second) {
-    it->second = FindSet(it->second);
+    return FindSet(it->second);
```
[the remark this answers is not in the document]
Comment (sebastianklose) [reply cartographer/mapping/internal/connected_components.cc]: I see. I will revert this and add a comment about path compression s.th. we don't try to remove it again.

**Comment by sebastianklose** [comment]
@wohe I don't see this as a matter of taste but rather code health, proper use of language features, readability and conforming to the standard.

And yes: if the standard and accompanied best practices change, I believe we should adapt it once we switch to the new standard.

Will look into the required changes in `cartographer::cloud` before submission.

**Comment by wohe** [comment]
@sebastianklose I was not commenting on whether this is a matter of taste, but that you present it as one. Calling it "sanity" or saying your change would be "proper use" is not an objective argument. I was asking that you provide the rationale why you think this change should be done. If the change is not subjective, you obviously should be able to do this. The benefit is that in the future, it is clear why this change was done.

**Comment by sebastianklose** [comment]
@wohe Sorry, I misinterpreted your comment. I did rephrase the description to be a bit more generic and changed the PR title.

**Pull request merged**